orangecontrib/bioinformatics/annotation/annotate_samples.py

```diff
@@ -0,0 +1,202 @@
+
+    Example of use:
+
+    >>> from Orange.data import Table
+    >>> from orangecontrib.bioinformatics.utils import serverfiles
+    >>> from orangecontrib.bioinformatics.annotation.annotate_samples import \
+    ...     AnnotateSamples
+    >>> from orangecontrib.bioinformatics.widgets.utils.data import TAX_ID
+    >>>
+    >>> data = Table("https://datasets.orange.biolab.si/sc/aml-1k.tab.gz")
+    >>> data.attributes[TAX_ID] = "9606"  # table needs to have an organism ID
+    >>> markers_path = serverfiles.localpath_download(
+    ...     'marker_genes','panglao_gene_markers.tab')
+    >>> marker = Table(markers_path)
+    >>>
+    >>> # filter only human markers
+    >>> from Orange.data.filter import FilterString, Values
+    >>> f = FilterString("Organism", FilterString.Equal, "Human")
+    >>> markers = Values([f])(markers)
+    >>>
+    >>> annotator = AnnotateSamples(p_value_th=0.05)
+    >>> annotations = annotator.annotate_samples(data, markers)
+
+    Attributes
+    ----------
+    p_value_th : float
+        A threshold for accepting the annotations. Annotations that has FDR
+        value bellow this threshold are used.
+    p_value_fun : callable, optional (defaults: statistics.Binomial().p_value)
+        A function that calculates p-value. It can be either
+        statistics.Binomial().p_value or hypergeom.sf.
+    """
+
+    def __init__(self, p_value_th, p_value_fun=statistics.Binomial().p_value):
+        if p_value_fun == hypergeom.sf:  # because sf accept x-1 instead of x
+            self.p_value_fun = lambda x, n, m, k: p_value_fun(x-1, n, m, k)
+        else:
+            self.p_value_fun = p_value_fun
+
+        self.p_threshold = p_value_th
+
+    @staticmethod
+    def _select_genes(data):
+        """
+        Function selects "over"-expressed attributes for items with Mann-Whitney
+        U test.
+
+        Parameters
+        ----------
+        data : Orange.data.Table
+            Tabular data
+
+        Returns
+        -------
+        :obj:`list`
+            Sets of selected attributes for each cell
+        """
+        if len(data.X) <= 1:
+            return [], []
+        # rank data
+        data_ge_ranked = rankdata(data.X, axis=0)
+
+        # compute U, mu, sigma
+        n = data_ge_ranked.shape[0]
+        n2 = n - 1
+        u = data_ge_ranked - 1
+        mu = n2 / 2
+        sigma = np.zeros(data_ge_ranked.shape[1])
+        for i in range(data_ge_ranked.shape[1]):
+            _, counts = np.unique(data_ge_ranked[:, i], return_counts=True)
+            sigma[i] = np.sqrt(
+                1 * n2 / 12 * ((n + 1) - np.sum((counts ** 3 - counts)) /
+                               (n * (n - 1))))
+
+        # compute z
+        z = (u - mu) / (sigma + 1e-16)
+
+        # gene selection
+        attributes_np = np.array([
+            a.attributes.get("Entrez ID") for a in data.domain.attributes])
+        attributes_sets = [set(attributes_np[row > 1]) - {None} for row in z]
```

Comment:
Do not hardcode a z-score threshold (of 1), but make it an argument to the function with default value 1. Pass it also as a parameter to `__init__` and use in `annotate_samples`. This way, this function can be used in stand-alone or within-class mode.